In [994]:
import pandas as pd

In [995]:
# Read uncleaned disney_movies.csv file
df_1 = pd.read_csv(r'C:\Users\ashwi\Documents\Data Analytics\QA\Data\Cleaned\Python\df_1.csv', index_col=False)

In [996]:
df_1.shape

(525, 21)

In [997]:
df_1.dtypes

Unnamed: 0            int64
movie_name           object
Directed by          object
Story by             object
Based on             object
Produced by          object
Starring             object
Music by             object
Productioncompany    object
Distributed by       object
Running time         object
Country              object
Language             object
Budget               object
Box office           object
Narrated by          object
Cinematography       object
Release date         object
Written by           object
Edited by            object
Screenplay by        object
dtype: object

In [998]:
list = df_1.columns.tolist()
sorted_list = sorted(list)
sorted_list

['Based on',
 'Box office',
 'Budget',
 'Cinematography',
 'Country',
 'Directed by',
 'Distributed by',
 'Edited by',
 'Language',
 'Music by',
 'Narrated by',
 'Produced by',
 'Productioncompany',
 'Release date',
 'Running time',
 'Screenplay by',
 'Starring',
 'Story by',
 'Unnamed: 0',
 'Written by',
 'movie_name']

In [999]:
df_1.head(1)

Unnamed: 0                                   movie_name  \
0           0  Snow White and the Seven Dwarfs (1937 film)   

                                                                          Directed by  \
0  ['David Hand', 'Perce Pearce', 'Larry Morey', 'Wilfred Jackson', 'Ben Sharpsteen']   

                                         Story by  \
0  ['Ted Sears', 'Earl Hurd', 'Merrill De Maris']   

                           Based on      Produced by  \
0  ['Snow White', 'Brothers Grimm']  ['Walt Disney']   

                                                                                                                 Starring  \
0  ['Adriana Caselotti', 'Roy Atwell', 'Pinto Colvig', 'Otis Harlan', 'Scotty Mattraw', 'Billy Gilbert', 'Eddie Collins']   

                                             Music by  \
0  ['Frank Churchill', 'Leigh Harline', 'Paul Smith']   

             Productioncompany          Distributed by  ...  \
0  ['Walt Disney Productions']  ['RKO Radio Pictures']  ...   

             Country     Language               Budget           Box office  \
0  ['United States']  ['English']  ['$1.5 million[1]']  ['$418 million[2]']   

  Narrated by Cinematography                Release date Written by Edited by  \
0         NaN            NaN  ['Carthay Circle Theatre']        NaN       NaN   

  Screenplay by  
0           NaN  

[1 rows x 21 columns]

## Remove leading [' and '] from columns

In [1000]:
columns_to_rm_brack = ['Based on', 'Box office', 'Budget', 'Cinematography', 'Country', 'Directed by', 'Distributed by', 'Edited by', 'Language', 'Music by', 'Narrated by', 'Produced by', 'Productioncompany', 'Release date', 'Running time', 'Screenplay by', 'Starring', 'Story by', 'Unnamed: 0', 'Written by', 'movie_name']

# function
def clean_column(value):
    if isinstance(value, str) and value.startswith("['") and value.endswith("']"):
        return value[2:-2]
    return value

# apply fuctnion to the specified columns
for column in columns_to_rm_brack:
    df_1[column] = df_1[column].apply(clean_column)

#display cleaned df
df_1.head(1)

Unnamed: 0                                   movie_name  \
0           0  Snow White and the Seven Dwarfs (1937 film)   

                                                                      Directed by  \
0  David Hand', 'Perce Pearce', 'Larry Morey', 'Wilfred Jackson', 'Ben Sharpsteen   

                                     Story by                      Based on  \
0  Ted Sears', 'Earl Hurd', 'Merrill De Maris  Snow White', 'Brothers Grimm   

   Produced by  \
0  Walt Disney   

                                                                                                             Starring  \
0  Adriana Caselotti', 'Roy Atwell', 'Pinto Colvig', 'Otis Harlan', 'Scotty Mattraw', 'Billy Gilbert', 'Eddie Collins   

                                         Music by        Productioncompany  \
0  Frank Churchill', 'Leigh Harline', 'Paul Smith  Walt Disney Productions   

       Distributed by  ...        Country Language           Budget  \
0  RKO Radio Pictures  ...  United States  English  $1.5 million[1]   

        Box office Narrated by Cinematography            Release date  \
0  $418 million[2]         NaN            NaN  Carthay Circle Theatre   

  Written by Edited by Screenplay by  
0        NaN       NaN           NaN  

[1 rows x 21 columns]

## Remove single and double quotes except apostrphes

In [1001]:
pd.set_option('display.max_colwidth', None)  # Remove the column width limit

df_1['Release date']

0                                                                                                                                                                                                                                                  Carthay Circle Theatre
1      Center Theatre', 'February\xa07,\xa01940\xa0(1940-02-07) (Center Theatre)[2]\nFebruary\xa023,\xa01940\xa0(1940-02-23) (United States)[3]', 'February\xa07,\xa01940\xa0(1940-02-07) (Center Theatre)[2]\nFebruary\xa023,\xa01940\xa0(1940-02-23) (United States)[3]
2                                                                                                                                                                                                                             \nNovember\xa013,\xa01940\xa0(1940-11-13)\n
3                                                                                                                                                                                                         

In [1002]:
# List of columns to process
columns_to_process = ['Based on', 'Box office', 'Budget', 'Cinematography', 'Country', 
                      'Directed by', 'Distributed by', 'Edited by', 'Language', 'Music by', 
                      'Narrated by', 'Produced by', 'Productioncompany', 'Release date', 
                      'Running time', 'Screenplay by', 'Starring', 'Story by', 
                      'Written by', 'movie_name']

# Removing single and double quotes from the specified columns
for column in columns_to_process:
    df_1[column] = df_1[column].str.replace(r"[\"']", "", regex=True)

df_1.head(1)

Unnamed: 0                                   movie_name  \
0           0  Snow White and the Seven Dwarfs (1937 film)   

                                                              Directed by  \
0  David Hand, Perce Pearce, Larry Morey, Wilfred Jackson, Ben Sharpsteen   

                                 Story by                    Based on  \
0  Ted Sears, Earl Hurd, Merrill De Maris  Snow White, Brothers Grimm   

   Produced by  \
0  Walt Disney   

                                                                                                 Starring  \
0  Adriana Caselotti, Roy Atwell, Pinto Colvig, Otis Harlan, Scotty Mattraw, Billy Gilbert, Eddie Collins   

                                     Music by        Productioncompany  \
0  Frank Churchill, Leigh Harline, Paul Smith  Walt Disney Productions   

       Distributed by  ...        Country Language           Budget  \
0  RKO Radio Pictures  ...  United States  English  $1.5 million[1]   

        Box office Narrated by Cinematography            Release date  \
0  $418 million[2]         NaN            NaN  Carthay Circle Theatre   

  Written by Edited by Screenplay by  
0        NaN       NaN           NaN  

[1 rows x 21 columns]

## Remove [#]

In [1003]:
df_1.columns

Index(['Unnamed: 0', 'movie_name', 'Directed by', 'Story by', 'Based on',
       'Produced by', 'Starring', 'Music by', 'Productioncompany',
       'Distributed by', 'Running time', 'Country', 'Language', 'Budget',
       'Box office', 'Narrated by', 'Cinematography', 'Release date',
       'Written by', 'Edited by', 'Screenplay by'],
      dtype='object')

In [1004]:
# function
def remove_square_brackets(value):
    import re
    return re.sub(r'\[\d+\]$', '', str(value))

# apply fuction to columns
df_1['Box office'] = df_1['Box office'].apply(remove_square_brackets)
df_1['Budget'] = df_1['Budget'].apply(remove_square_brackets)
df_1['Release date'] = df_1['Release date'].apply(remove_square_brackets)

df_1.head(1)

Unnamed: 0                                   movie_name  \
0           0  Snow White and the Seven Dwarfs (1937 film)   

                                                              Directed by  \
0  David Hand, Perce Pearce, Larry Morey, Wilfred Jackson, Ben Sharpsteen   

                                 Story by                    Based on  \
0  Ted Sears, Earl Hurd, Merrill De Maris  Snow White, Brothers Grimm   

   Produced by  \
0  Walt Disney   

                                                                                                 Starring  \
0  Adriana Caselotti, Roy Atwell, Pinto Colvig, Otis Harlan, Scotty Mattraw, Billy Gilbert, Eddie Collins   

                                     Music by        Productioncompany  \
0  Frank Churchill, Leigh Harline, Paul Smith  Walt Disney Productions   

       Distributed by  ...        Country Language        Budget  \
0  RKO Radio Pictures  ...  United States  English  $1.5 million   

     Box office Narrated by Cinematography            Release date Written by  \
0  $418 million         NaN            NaN  Carthay Circle Theatre        NaN   

  Edited by Screenplay by  
0       NaN           NaN  

[1 rows x 21 columns]

## Null, Missing values, duplicates, datatypes

### Inspect the Data

In [1005]:
# Display first few rows
df_1.head(1)

Unnamed: 0                                   movie_name  \
0           0  Snow White and the Seven Dwarfs (1937 film)   

                                                              Directed by  \
0  David Hand, Perce Pearce, Larry Morey, Wilfred Jackson, Ben Sharpsteen   

                                 Story by                    Based on  \
0  Ted Sears, Earl Hurd, Merrill De Maris  Snow White, Brothers Grimm   

   Produced by  \
0  Walt Disney   

                                                                                                 Starring  \
0  Adriana Caselotti, Roy Atwell, Pinto Colvig, Otis Harlan, Scotty Mattraw, Billy Gilbert, Eddie Collins   

                                     Music by        Productioncompany  \
0  Frank Churchill, Leigh Harline, Paul Smith  Walt Disney Productions   

       Distributed by  ...        Country Language        Budget  \
0  RKO Radio Pictures  ...  United States  English  $1.5 million   

     Box office Narrated by Cinematography            Release date Written by  \
0  $418 million         NaN            NaN  Carthay Circle Theatre        NaN   

  Edited by Screenplay by  
0       NaN           NaN  

[1 rows x 21 columns]

In [1006]:
# Attain summary of DataFrame (dF)

In [1007]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525 entries, 0 to 524
Data columns (total 21 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Unnamed: 0         525 non-null    int64 
 1   movie_name         525 non-null    object
 2   Directed by        525 non-null    object
 3   Story by           169 non-null    object
 4   Based on           290 non-null    object
 5   Produced by        517 non-null    object
 6   Starring           493 non-null    object
 7   Music by           518 non-null    object
 8   Productioncompany  525 non-null    object
 9   Distributed by     525 non-null    object
 10  Running time       516 non-null    object
 11  Country            521 non-null    object
 12  Language           502 non-null    object
 13  Budget             525 non-null    object
 14  Box office         525 non-null    object
 15  Narrated by        59 non-null     object
 16  Cinematography     399 non-null    object
 1

In [1008]:
# attain summary statistics for numeric columns
df_1.describe()

Unnamed: 0
count  525.000000
mean   262.000000
std    151.698715
min      0.000000
25%    131.000000
50%    262.000000
75%    393.000000
max    524.000000

### Missing Values

In [1009]:
missing_values = df_1.isnull().sum()
print("Missing values:\n", missing_values)

Missing values:
 Unnamed: 0             0
movie_name             0
Directed by            0
Story by             356
Based on             235
Produced by            8
Starring              32
Music by               7
Productioncompany      0
Distributed by         0
Running time           9
Country                4
Language              23
Budget                 0
Box office             0
Narrated by          466
Cinematography       126
Release date           0
Written by           301
Edited by             42
Screenplay by        263
dtype: int64


Can fill in missing values with mean, median, or drop rows; if cannot find suitable data from other columns

E.g., df_1.fillna(df_1.mean(), inplace=TRUE)

### Check for Duplicates

In [1010]:
# Working columns movie_name, running time, release date, box office

df_2 = df_1.loc[:, ['movie_name', 'Running time', 'Release date', 'Box office']]

In [1011]:
duplicates = df_2.duplicated().sum()
print('Number of duplicate rows:', duplicates)

Number of duplicate rows: 0


If duplicate rows found, removable:

df_1.drop_duplicates(inplace=True)

In [1012]:
df_2.dtypes

movie_name      object
Running time    object
Release date    object
Box office      object
dtype: object

### Validate Data Types

#### movie_name

In [1013]:
df_2['movie_name']

0      Snow White and the Seven Dwarfs (1937 film)
1                            Pinocchio (1940 film)
2                             Fantasia (1940 film)
3                 The Reluctant Dragon (1941 film)
4                                            Dumbo
                          ...                     
520                                    Elio (film)
521                          Out of My Mind (film)
522                        Young Woman and the Sea
523                  Lilo & Stitch (upcoming film)
524                                     Tron: Ares
Name: movie_name, Length: 525, dtype: object

In [1014]:
# function
def find_non_string_characters(s):
    non_string_characters = ''.join([char for char in s if not char.isalpha()])
    return non_string_characters

# apply fuction to column
non_string_characters_in_movie_names = df_2['movie_name'].apply(find_non_string_characters)

# print non-string characters
print("Non-string characters found in movie names:")
print(non_string_characters_in_movie_names[non_string_characters_in_movie_names != ""])

Non-string characters found in movie names:
0            (1937 )
1            (1940 )
2            (1940 )
3            (1941 )
6                   
           ...      
520               ()
521               ()
522                 
523           &  ( )
524               : 
Name: movie_name, Length: 512, dtype: object


In [1015]:
df_2.columns

Index(['movie_name', 'Running time', 'Release date', 'Box office'], dtype='object')

In [1016]:
# remove non-string characters and empty spaces

# remove & and :
df_2['movie_name'] = df_2['movie_name'].str.replace(r'[&:]', '', regex=True)

# remove empty spaces after four numericl characters
df_2['movie_name'] = df_2['movie_name'].str.replace(r'(\d{4})\s', r'\1', regex=True)

# remove () when othing inside
df_2['movie_name'] = df_2['movie_name'].str.replace(r'\(\s*\)', '', regex=True)

In [1017]:
df_2.columns

Index(['movie_name', 'Running time', 'Release date', 'Box office'], dtype='object')

In [1018]:
# function
def find_non_string_characters(s):
    non_string_characters = ''.join([char for char in s if not char.isalpha()])
    return non_string_characters

# apply fuction to column
non_string_characters_in_movie_names = df_2['movie_name'].apply(find_non_string_characters)

# print non-string characters
print("Non-string characters found in movie names:")
print(non_string_characters_in_movie_names[non_string_characters_in_movie_names != ""])

Non-string characters found in movie names:
0            (1937)
1            (1940)
2            (1940)
3            (1941)
6                  
           ...     
520              ()
521              ()
522                
523             ( )
524                
Name: movie_name, Length: 512, dtype: object


In [1019]:
df_2.columns

Index(['movie_name', 'Running time', 'Release date', 'Box office'], dtype='object')

In [1020]:
# Remove parentheses with no characters inside from 'movie_name' column
df_2['movie_name'] = df_2['movie_name'].str.replace(r'\s*\(\s*\)', '', regex=True)

# Display the updated DataFrame
df_2.head(1)

movie_name Running time  \
0  Snow White and the Seven Dwarfs (1937film)   83 minutes   

             Release date    Box office  
0  Carthay Circle Theatre  $418 million

In [1021]:
# Remove parentheses with no characters inside from 'movie_name' column
df_2['movie_name'] = df_2['movie_name'].str.replace(r'\s*\(\s*\)', '', regex=True)

# Display the updated DataFrame
print(df_2.head(1))

                                   movie_name Running time  \
0  Snow White and the Seven Dwarfs (1937film)   83 minutes   

             Release date    Box office  
0  Carthay Circle Theatre  $418 million  


In [1022]:
# Remove parentheses with no characters inside and any spaces
df_2['movie_name'] = df_2['movie_name'].str.replace(r'\s*\(\s*\)', '', regex=True)

In [1023]:
df_2.head(1)

movie_name Running time  \
0  Snow White and the Seven Dwarfs (1937film)   83 minutes   

             Release date    Box office  
0  Carthay Circle Theatre  $418 million

In [1024]:
df_2['movie_name'] = df_2['movie_name'].astype(str)

In [1025]:
df_2.dtypes

movie_name      object
Running time    object
Release date    object
Box office      object
dtype: object

In [1026]:
df_2.columns

Index(['movie_name', 'Running time', 'Release date', 'Box office'], dtype='object')

#### Running time

In [1027]:
df_2['Running time']

0          83 minutes
1          88 minutes
2      126 minutes[1]
3          74 minutes
4          64 minutes
            ...      
520               NaN
521    102 minutes[1]
522               NaN
523               NaN
524               NaN
Name: Running time, Length: 525, dtype: object

In [1028]:
# change column name
df_2 = df_2.rename(columns = {'Running time': 'Running time (min)'})

In [1029]:
df_2.columns

Index(['movie_name', 'Running time (min)', 'Release date', 'Box office'], dtype='object')

In [1030]:
import re

In [1031]:
non_alphanumeric_values = df_2['Running time (min)'][df_2['Running time (min)'].str.contains(r'[^a-zA-Z0-9\s]', na=False)]

print(non_alphanumeric_values)

2                                                        126 minutes[1]
10                                                        94 minutes[1]
23     60 minutes (VHS and Wild Discovery version)71 minutes (original)
54                                                       105 minutes[1]
65                                                              80 min.
                                     ...                               
510                                                      101 minutes[1]
512                                                      154 minutes[3]
513                                                      123 minutes[1]
515                                                       95 minutes[1]
521                                                      102 minutes[1]
Name: Running time (min), Length: 162, dtype: object


In [1032]:
df_2.columns

Index(['movie_name', 'Running time (min)', 'Release date', 'Box office'], dtype='object')

In [1033]:
# function
def remove_square_brackets(value):
    if isinstance(value, str):
        return re.sub(r'\[\d+\]', '', value)
    else:
        return value

# apply fucntion to column
df_2['Running time (min)'] = df_2['Running time (min)'].apply(remove_square_brackets)

# display modified dataFrame
df_2['Running time (min)']

0       83 minutes
1       88 minutes
2      126 minutes
3       74 minutes
4       64 minutes
          ...     
520            NaN
521    102 minutes
522            NaN
523            NaN
524            NaN
Name: Running time (min), Length: 525, dtype: object

In [1034]:
# Display values from 'Running time (min)' that do not match the pattern
pattern = r'^\d+\s*(min|minutes)$'
non_matching_values = df_2[~df_2['Running time (min)'].str.match(pattern, na=False)]
non_matching_values.head(1)

movie_name  \
23  The Vanishing Prairie   

                                                  Running time (min)  \
23  60 minutes (VHS and Wild Discovery version)71 minutes (original)   

                                 Release date  \
23  \nAugust\xa017,\xa01954\xa0(1954-08-17)\n   

                                 Box office  
23  $1.75 million (US and Canadian rentals)

In [1035]:
# Regular expression pattern to match values with numbers followed by 'min', 'minutes', or 'hours'
pattern = r'\d+\s*(min|minutes|hours)'

# Filter the DataFrame to keep only the values that match the pattern
matching_values = df_2[df_2['Running time (min)'].str.contains(pattern, case=False, na=False)]

# Display the filtered DataFrame
print(matching_values[['Running time (min)']])

    Running time (min)
0           83 minutes
1           88 minutes
2          126 minutes
3           74 minutes
4           64 minutes
..                 ...
513        123 minutes
514         90 minutes
515         95 minutes
516         62 minutes
521        102 minutes

[515 rows x 1 columns]


C:\Users\ashwi\AppData\Local\Temp\ipykernel_4744\3848823176.py:5: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  matching_values = df_2[df_2['Running time (min)'].str.contains(pattern, case=False, na=False)]


In [1036]:
# Display values from 'Running time (min)' that do not match the pattern
pattern = r'^\d+\s*(min|minutes)$'
non_matching_values = df_2[~df_2['Running time (min)'].str.match(pattern, na=False)]
non_matching_values.head()

movie_name  \
23               The Vanishing Prairie   
65       The Incredible Journey (film)   
72                     Those Calloways   
85            The Happiest Millionaire   
89  The Horse in the Gray Flannel Suit   

                                                  Running time (min)  \
23  60 minutes (VHS and Wild Discovery version)71 minutes (original)   
65                                                           80 min.   
72                                                         131 mins.   
85                       [Los Angeles, New York City, Directors cut]   
89                                                         113 mins.   

                                   Release date  \
23    \nAugust\xa017,\xa01954\xa0(1954-08-17)\n   
65      November\xa020,\xa01963\xa0(1963-11-20)   
72                            November 10, 1964   
85                        Radio City Music Hall   
89  \nDecember\xa020,\xa01968\xa0(1968-12-20)\n   

                                 Box office  
23  $1.75 million (US and Canadian rentals)  
65       $4.2 million (US/Canada) (rentals)  
72          $3,500,000 (US/ Canada rentals)  
85         $5 million (U.S./Canada rentals)  
89       $3.3 million  (US/ Canada rentals)

In [1037]:
# Update value in row 23 to '60 min'
df_2.iloc[23, df_2.columns.get_loc('Running time (min)')] = '60 min'

# Update value in row 85 to empty string
df_2.iloc[85, df_2.columns.get_loc('Running time (min)')] = ''

# Display the updated DataFrame
print(df_2[['Running time (min)']])

    Running time (min)
0           83 minutes
1           88 minutes
2          126 minutes
3           74 minutes
4           64 minutes
..                 ...
520                NaN
521        102 minutes
522                NaN
523                NaN
524                NaN

[525 rows x 1 columns]


In [1038]:
# Display values from 'Running time (min)' that do not match the pattern
pattern = r'^\d+\s*(min|minutes)$'
non_matching_values = df_2[~df_2['Running time (min)'].str.match(pattern, na=False)]
non_matching_values.head()

movie_name  \
65        The Incredible Journey (film)   
72                      Those Calloways   
85             The Happiest Millionaire   
89   The Horse in the Gray Flannel Suit   
101            Bedknobs and Broomsticks   

                                                               Running time (min)  \
65                                                                        80 min.   
72                                                                      131 mins.   
85                                                                                  
89                                                                      113 mins.   
101  118 minutes (1971 original version)139 minutes (1996 reconstruction version)   

                                                                                                            Release date  \
65                                                                               November\xa020,\xa01963\xa0(1963-11-20)   
72                                                                                                     November 10, 1964   
85                                                                                                 Radio City Music Hall   
89                                                                           \nDecember\xa020,\xa01968\xa0(1968-12-20)\n   
101  \nOctober\xa07,\xa01971\xa0(1971-10-07) (United Kingdom)\nDecember\xa013,\xa01971\xa0(1971-12-13) (United States)\n   

                             Box office  
65   $4.2 million (US/Canada) (rentals)  
72      $3,500,000 (US/ Canada rentals)  
85     $5 million (U.S./Canada rentals)  
89   $3.3 million  (US/ Canada rentals)  
101                    $17.9\xa0million

Row 101 shows 118 min for 1971 film and 139 min for 1996 film
Release date shows 1971 film.

In [1039]:
# Update value in row 101 to '118 min'
df_2.iloc[101, df_2.columns.get_loc('Running time (min)')] = '118 min'

# Display the updated DataFrame
print(df_2[['Running time (min)']])

    Running time (min)
0           83 minutes
1           88 minutes
2          126 minutes
3           74 minutes
4           64 minutes
..                 ...
520                NaN
521        102 minutes
522                NaN
523                NaN
524                NaN

[525 rows x 1 columns]


In [1040]:
# Display values from 'Running time (min)' that do not match the pattern
pattern = r'^\d+\s*(min|minutes)$'
non_matching_values = df_2[~df_2['Running time (min)'].str.match(pattern, na=False)]
non_matching_values.head()

movie_name Running time (min)  \
65        The Incredible Journey (film)            80 min.   
72                      Those Calloways          131 mins.   
85             The Happiest Millionaire                      
89   The Horse in the Gray Flannel Suit          113 mins.   
109            One Little Indian (film)         90 Minutes   

                                    Release date  \
65       November\xa020,\xa01963\xa0(1963-11-20)   
72                             November 10, 1964   
85                         Radio City Music Hall   
89   \nDecember\xa020,\xa01968\xa0(1968-12-20)\n   
109      \nJune\xa020,\xa01973\xa0(1973-06-20)\n   

                             Box office  
65   $4.2 million (US/Canada) (rentals)  
72      $3,500,000 (US/ Canada rentals)  
85     $5 million (U.S./Canada rentals)  
89   $3.3 million  (US/ Canada rentals)  
109                          $2 million

In [1041]:
# Extract numbers from the 'Running time (min)' column
df_2['Running time (min)'] = df_2['Running time (min)'].str.extract(r'(\d+)')

# Display the updated DataFrame
print(df_2[['Running time (min)']])

    Running time (min)
0                   83
1                   88
2                  126
3                   74
4                   64
..                 ...
520                NaN
521                102
522                NaN
523                NaN
524                NaN

[525 rows x 1 columns]


In [1042]:
df_2.head(1)

movie_name Running time (min)  \
0  Snow White and the Seven Dwarfs (1937film)                 83   

             Release date    Box office  
0  Carthay Circle Theatre  $418 million

In [1043]:
df_2.dtypes

movie_name            object
Running time (min)    object
Release date          object
Box office            object
dtype: object

In [1044]:


# Assuming df_2 is your DataFrame with the 'Running time (min)' column

# Convert the 'Running time (min)' column to timedelta format
df_2['Running time (min)'] = pd.to_timedelta(df_2['Running time (min)'] + ' minutes')

# Display the updated DataFrame
print(df_2[['Running time (min)']])

    Running time (min)
0      0 days 01:23:00
1      0 days 01:28:00
2      0 days 02:06:00
3      0 days 01:14:00
4      0 days 01:04:00
..                 ...
520                NaT
521    0 days 01:42:00
522                NaT
523                NaT
524                NaT

[525 rows x 1 columns]


In [1045]:
df_2.dtypes

movie_name                     object
Running time (min)    timedelta64[ns]
Release date                   object
Box office                     object
dtype: object

In [1046]:
# Convert timedelta values to string representation
df_2['Running time (min)'] = df_2['Running time (min)'].astype(str)

# Remove days and milliseconds from the string representation
df_2['Running time (min)'] = df_2['Running time (min)'].str.extract(r'(\d+:\d+)')

# Display the updated DataFrame
df_2.head(1)

movie_name Running time (min)  \
0  Snow White and the Seven Dwarfs (1937film)              01:23   

             Release date    Box office  
0  Carthay Circle Theatre  $418 million

#### Release date

In [1047]:
df_2.columns

Index(['movie_name', 'Running time (min)', 'Release date', 'Box office'], dtype='object')

In [1048]:
df_2.head(2)

movie_name Running time (min)  \
0  Snow White and the Seven Dwarfs (1937film)              01:23   
1                        Pinocchio (1940film)              01:28   

                                                                                                                                                                                                                                                  Release date  \
0                                                                                                                                                                                                                                       Carthay Circle Theatre   
1  Center Theatre, February\xa07,\xa01940\xa0(1940-02-07) (Center Theatre)[2]\nFebruary\xa023,\xa01940\xa0(1940-02-23) (United States)[3], February\xa07,\xa01940\xa0(1940-02-07) (Center Theatre)[2]\nFebruary\xa023,\xa01940\xa0(1940-02-23) (United States)   

        Box office  
0     $418 million  
1  $164\xa0million

In [1049]:
# Assuming df_2 is your DataFrame with the specified columns

# Filter rows where 'Release date' does not match the specified patterns
pattern = r'(\d{2}-\d{2}-\d{4})|(\d{4}-\d{2}-\d{2})|(\d{8})'
non_matching_dates = df_2[~df_2['Release date'].str.match(pattern, na=False)]

# Display the filtered DataFrame
non_matching_dates

movie_name Running time (min)  \
0    Snow White and the Seven Dwarfs (1937film)              01:23   
1                          Pinocchio (1940film)              01:28   
2                           Fantasia (1940film)              02:06   
3               The Reluctant Dragon (1941film)              01:14   
4                                         Dumbo              01:04   
..                                          ...                ...   
520                                 Elio (film)                NaN   
521                       Out of My Mind (film)              01:42   
522                     Young Woman and the Sea                NaN   
523                Lilo  Stitch (upcoming film)                NaN   
524                                   Tron Ares                NaN   

                                                                                                                                                                                                                                                    Release date  \
0                                                                                                                                                                                                                                         Carthay Circle Theatre   
1    Center Theatre, February\xa07,\xa01940\xa0(1940-02-07) (Center Theatre)[2]\nFebruary\xa023,\xa01940\xa0(1940-02-23) (United States)[3], February\xa07,\xa01940\xa0(1940-02-07) (Center Theatre)[2]\nFebruary\xa023,\xa01940\xa0(1940-02-23) (United States)   
2                                                                                                                                                                                                                    \nNovember\xa013,\xa01940\xa0(1940-11-13)\n   
3                                                                                                                                                                                                                            June\xa027,\xa01941\xa0(1941-06-27)   
4                                                                                                                                                       October\xa023,\xa01941\xa0(1941-10-23) (New York City)[1]\nOctober\xa031,\xa01941\xa0(1941-10-31) (U.S.)   
..                                                                                                                                                                                                                                                           ...   
520                                                                                                                                                                                                                      \nJune\xa013,\xa02025\xa0(2025-06-13)\n   
521                                                                                                                                                                                                                                                     Sundance   
522                                                                                                                                                                                                                                                          nan   
523                                                                                                                                                                                                                                                          nan   
524                                                                                                                                                                                                                                                          nan   

                                                                           

In [1050]:
# Update release date in row 1
df_2.at[1, 'Release date'] = '1940-02-07'

In [1051]:
# Assuming df_2 is your DataFrame with the specified columns

# Filter rows where 'Release date' does not match the specified patterns
pattern = r'(\d{2}-\d{2}-\d{4})|(\d{4}-\d{2}-\d{2})|(\d{8})'
non_matching_dates = df_2[~df_2['Release date'].str.match(pattern, na=False)]

# Display the filtered DataFrame
non_matching_dates

movie_name Running time (min)  \
0    Snow White and the Seven Dwarfs (1937film)              01:23   
2                           Fantasia (1940film)              02:06   
3               The Reluctant Dragon (1941film)              01:14   
4                                         Dumbo              01:04   
5                                         Bambi              01:10   
..                                          ...                ...   
520                                 Elio (film)                NaN   
521                       Out of My Mind (film)              01:42   
522                     Young Woman and the Sea                NaN   
523                Lilo  Stitch (upcoming film)                NaN   
524                                   Tron Ares                NaN   

                                                                                                     Release date  \
0                                                                                          Carthay Circle Theatre   
2                                                                     \nNovember\xa013,\xa01940\xa0(1940-11-13)\n   
3                                                                             June\xa027,\xa01941\xa0(1941-06-27)   
4        October\xa023,\xa01941\xa0(1941-10-23) (New York City)[1]\nOctober\xa031,\xa01941\xa0(1941-10-31) (U.S.)   
5    London, August\xa09,\xa01942\xa0(1942-08-09) (London)\nAugust\xa021,\xa01942\xa0(1942-08-21) (United States)   
..                                                                                                            ...   
520                                                                       \nJune\xa013,\xa02025\xa0(2025-06-13)\n   
521                                                                                                      Sundance   
522                                                                                                           nan   
523                                                                                                           nan   
524                                                                                                           nan   

                                                                                                  Box office  
0                                                                                               $418 million  
2    $76.4–$83.3 million (United States and Canada)[4][5], $76.4–$83.3 million (United States and Canada)[4]  
3                                                                              $960,000 (worldwide rentals)   
4                                                    >$1.3 million (est. United States/Canada rentals, 1941)  
5                                                                                             $267.4 million  
..                                                                                                       ...  
520                                                                                                      nan  
521                                                                                                      nan  
522                                                                                                      nan  
523                                                                                                      nan  
524                                                                                                      nan  

[524 rows x 4 columns]

In [1052]:
# Extract the first date formatted as ####-##-## from 'Release date' column for each row
df_2['Release date'] = df_2['Release date'].str.extract(r'(\d{4}-\d{2}-\d{2})')

# Display the updated DataFrame
df_2

movie_name Running time (min)  \
0    Snow White and the Seven Dwarfs (1937film)              01:23   
1                          Pinocchio (1940film)              01:28   
2                           Fantasia (1940film)              02:06   
3               The Reluctant Dragon (1941film)              01:14   
4                                         Dumbo              01:04   
..                                          ...                ...   
520                                 Elio (film)                NaN   
521                       Out of My Mind (film)              01:42   
522                     Young Woman and the Sea                NaN   
523                Lilo  Stitch (upcoming film)                NaN   
524                                   Tron Ares                NaN   

    Release date  \
0            NaN   
1     1940-02-07   
2     1940-11-13   
3     1941-06-27   
4     1941-10-23   
..           ...   
520   2025-06-13   
521          NaN   
522          NaN   
523          NaN   
524          NaN   

                                                                                                  Box office  
0                                                                                               $418 million  
1                                                                                            $164\xa0million  
2    $76.4–$83.3 million (United States and Canada)[4][5], $76.4–$83.3 million (United States and Canada)[4]  
3                                                                              $960,000 (worldwide rentals)   
4                                                    >$1.3 million (est. United States/Canada rentals, 1941)  
..                                                                                                       ...  
520                                                                                                      nan  
521                                                                                                      nan  
522                                                                                                      nan  
523                                                                                                      nan  
524                                                                                                      nan  

[525 rows x 4 columns]

In [1053]:
# Convert 'Release date' column to date datatype in YYYY-MM-DD format
df_2['Release date'] = pd.to_datetime(df_2['Release date'], format='%Y-%m-%d')

# Display the updated DataFrame
print(df_2['Release date'])

0            NaT
1     1940-02-07
2     1940-11-13
3     1941-06-27
4     1941-10-23
         ...    
520   2025-06-13
521          NaT
522          NaT
523          NaT
524          NaT
Name: Release date, Length: 525, dtype: datetime64[ns]


In [1054]:
df_2

movie_name Running time (min)  \
0    Snow White and the Seven Dwarfs (1937film)              01:23   
1                          Pinocchio (1940film)              01:28   
2                           Fantasia (1940film)              02:06   
3               The Reluctant Dragon (1941film)              01:14   
4                                         Dumbo              01:04   
..                                          ...                ...   
520                                 Elio (film)                NaN   
521                       Out of My Mind (film)              01:42   
522                     Young Woman and the Sea                NaN   
523                Lilo  Stitch (upcoming film)                NaN   
524                                   Tron Ares                NaN   

    Release date  \
0            NaT   
1     1940-02-07   
2     1940-11-13   
3     1941-06-27   
4     1941-10-23   
..           ...   
520   2025-06-13   
521          NaT   
522          NaT   
523          NaT   
524          NaT   

                                                                                                  Box office  
0                                                                                               $418 million  
1                                                                                            $164\xa0million  
2    $76.4–$83.3 million (United States and Canada)[4][5], $76.4–$83.3 million (United States and Canada)[4]  
3                                                                              $960,000 (worldwide rentals)   
4                                                    >$1.3 million (est. United States/Canada rentals, 1941)  
..                                                                                                       ...  
520                                                                                                      nan  
521                                                                                                      nan  
522                                                                                                      nan  
523                                                                                                      nan  
524                                                                                                      nan  

[525 rows x 4 columns]

In [1055]:
df_2['Release date'].max()

Timestamp('2025-06-13 00:00:00')

Outlier Detected

In [1056]:
## Limited max date to 2024-05-31

# Convert 'Release date' column to datetime datatype if it's not already in that format
df_2['Release date'] = pd.to_datetime(df_2['Release date'])

# Filter rows where 'Release date' > '2024-02-31'
filtered_df = df_2[df_2['Release date'] <= '2024-05-31']

# Display the filtered DataFrame
df_2_dated = filtered_df


In [1057]:
df_2_dated.head(1)

movie_name Running time (min) Release date       Box office
1  Pinocchio (1940film)              01:28   1940-02-07  $164\xa0million

In [1058]:
# Filter rows where 'Release date' is not in the YYYY-MM-DD format with numeric characters
pattern = r'^(?!20\d{2}-\d{2}-\d{2}$).*$'
non_numeric_dates = df_2_dated[df_2_dated['Release date'].astype(str).str.match(pattern)]

# Display the filtered DataFrame
non_numeric_dates

movie_name Running time (min) Release date  \
1               Pinocchio (1940film)              01:28   1940-02-07   
2                Fantasia (1940film)              02:06   1940-11-13   
3    The Reluctant Dragon (1941film)              01:14   1941-06-27   
4                              Dumbo              01:04   1941-10-23   
5                              Bambi              01:10   1942-08-09   
..                               ...                ...          ...   
240                Tarzan (1999film)              01:28   1999-06-12   
241          Inspector Gadget (film)              01:18   1999-07-23   
242            Hercules Zero to Hero              01:10   1999-08-17   
244    Mickeys Once Upon a Christmas              01:06   1999-11-09   
245                      Toy Story 2              01:32   1999-11-13   

                                                                                                  Box office  
1                                                                                            $164\xa0million  
2    $76.4–$83.3 million (United States and Canada)[4][5], $76.4–$83.3 million (United States and Canada)[4]  
3                                                                              $960,000 (worldwide rentals)   
4                                                    >$1.3 million (est. United States/Canada rentals, 1941)  
5                                                                                             $267.4 million  
..                                                                                                       ...  
240                                                                                           $448.2 million  
241                                                                                           $134.4 million  
242                                                                                                      nan  
244                                                                                                      nan  
245                                                            $511.4\xa0million[4][3], $511.4\xa0million[4]  

[216 rows x 4 columns]

In [1059]:
# Filter rows where the month is smaller than 01 or larger than 12
invalid_month_rows = df_2_dated[(df_2_dated['Release date'].dt.month < 1) | (df_2_dated['Release date'].dt.month > 12)]

# Display the filtered DataFrame
invalid_month_rows

Empty DataFrame
Columns: [movie_name, Running time (min), Release date, Box office]
Index: []

In [1061]:
# Filter rows where day is smaller than 01 or larger than 31
invalid_dates = df_2_dated[(df_2_dated['Release date'].dt.day < 1) | (df_2_dated['Release date'].dt.day > 31)]

# Display the filtered DataFrame
invalid_dates

Empty DataFrame
Columns: [movie_name, Running time (min), Release date, Box office]
Index: []

In [1065]:
df_2_dated

movie_name Running time (min) Release date  \
1               Pinocchio (1940film)              01:28   1940-02-07   
2                Fantasia (1940film)              02:06   1940-11-13   
3    The Reluctant Dragon (1941film)              01:14   1941-06-27   
4                              Dumbo              01:04   1941-10-23   
5                              Bambi              01:10   1942-08-09   
..                               ...                ...          ...   
506                   Chang Can Dunk              01:49   2023-03-10   
507                 Peter Pan  Wendy              01:49   2023-04-28   
508                    Crater (film)              01:45   2023-05-12   
511               Worlds Best (film)              01:41   2023-06-23   
514         Dashing Through the Snow              01:30   2023-11-17   

                                                                                                  Box office  
1                                                                                            $164\xa0million  
2    $76.4–$83.3 million (United States and Canada)[4][5], $76.4–$83.3 million (United States and Canada)[4]  
3                                                                              $960,000 (worldwide rentals)   
4                                                    >$1.3 million (est. United States/Canada rentals, 1941)  
5                                                                                             $267.4 million  
..                                                                                                       ...  
506                                                                                                      nan  
507                                                                                                      nan  
508                                                                                                      nan  
511                                                                                                      nan  
514                                                                                                      nan  

[371 rows x 4 columns]

#### Box office